uberenv.py

```diff
@@ -386,82 +389,44 @@ def config_dir(self):
         return spack_config_dir
 
 
-    def disable_spack_config_scopes(self,spack_dir):
-        # disables all config scopes except "defaults", which we will
-        # force our settings into
-        spack_lib_config = pjoin(spack_dir,"lib","spack","spack","config.py")
-        print("[disabling config scope (except defaults) in: {}]".format(spack_lib_config))
-        cfg_script = open(spack_lib_config).read()
-        for cfg_scope_stmt in ["('system', os.path.join(spack.paths.system_etc_path, 'spack')),",
-                            "('site', os.path.join(spack.paths.etc_path, 'spack')),",
-                            "('user', spack.paths.user_config_path)"]:
-            cfg_script = cfg_script.replace(cfg_scope_stmt,
-                                            "#DISABLED BY UBERENV: " + cfg_scope_stmt)
-        open(spack_lib_config,"w").write(cfg_script)
-
-
-    def patch(self):
-
-        cfg_dir = self.config_dir()
-        spack_dir = self.dest_spack
-
-        # force spack to use only "defaults" config scope
-        self.disable_spack_config_scopes(spack_dir)
-        spack_etc_defaults_dir = pjoin(spack_dir,"etc","spack","defaults")
+    def load(self):
+        # load spack environment, potentially overriding spack defaults
+        # uberenv used to handle defaults overriding, now spack environment can
+        # be used to do so, but it relies on each project to make sure their
+        # environment is well designed.
 
-        # copy in "defaults" config.yaml
-        config_yaml = os.path.abspath(pjoin(self.uberenv_path,"spack_configs","config.yaml"))
-        sexe("cp {} {}/".format(config_yaml, spack_etc_defaults_dir ), echo=True)
+        self.spack_cmd = "{} -e {}".format(self.spack_cmd,self.config_dir())
```

Comment:
Is there a reason you went with -e instead of loading the environment?  My concern is if you need to debug your build you would need to remember to add the -e to your spack commands.

Reply:
Unfortunately there is a reason:
I tried first with loading the environment, but loading from a script is not reliable/persistent in my experience. It seems to be because it would load the environment in a subprocess and then kill it (the subprocess).
I asked @becker33. The other solution would be to use "spack python" as an interpreter, but it’s not compatible with the fact uberenv clones spack.

Reply:
I think the "-e" method isn’t stupid though, the way I coded it I mean. Maybe we could make it more explicit what is the environment and the command to use to reproduce the behavior.

Like an output
"Uberenv: Now using spack environment `scripts/uberenv/spack-config/toss3...`"
"Ubernev: load environment with : "
"Uberenv: `> spack env activate scripts/uberenv/spack-config/toss3...`"
"Uberenv: or"
"Uberenv: `> spack -e scripts/uberenv/spack-config/toss3... <action>`"

Reply:
As long as there is a reason, this is the whole purpose of uberenv is to encapsulate what we feel is the best way to use spack for us.  If loading the environment doesn't work right now then we have to go with the -e option. Just make sure that it prints out the extra options so we can copy and paste the command.  Maybe when it errors out re-print the command to the screen so it is easy to find.

I would try to get them to fix the env command while moving forward with this solution though.

Reply:
What do you mean by "fix the env command"?

Reply:
I may not have been clear: It’s not spack fault, it’s because we are calling bash command in a python script. We cannot modify the calling shell environment, and creating a subprocess makes changes non-persistent.

Reply:
If you want to get really involved, we actually do this in Spack. But it requires dumping the environment of the subprocess to json and reading the json back in from the calling process, and using that to set the environment.

Would it be helpful to have a `--python` option (like the existing `--sh` and `--csh` options) to emit python code to set the loaded environment? We could probably add that reasonably easily, and then you could just take the output of the command and run `exec` on it.

Reply:
Thanks @becker33 for the answer. It seems interesting to me, but I wonder if it would be worth the effort. The way we do it now is that we change "spack_cmd" to be `spack -e path/to/env` as you suggested. And I think it does the job.

It seems to me that what you propose would be helpful if we wanted something like `uberenv activate`. However, I think it would just overlap with existing features in Spack, and we are trying to avoid that.

Reply:
Ok. Don't be shy about feature requests if there's something y'all need.

Reply:
can the env setup be persistent beyond the current shell instance?

That matches things like pip and conda, right?

Reply:
@cyrush I’m not sure I understand the question. 

> be persistent beyond the shell instance

Would you like the "activation" or the "configuration" to persist ?

Reply:
yes, our current setup modifies spack so that the compiler and package settings persist. 

Any further spack calls (driven by uberenv or manually) will respect those changes without having to know to invoke extra commands.

Reply:
Now I understand. And it’s true that it is not the case with the present PR.

What is the exact usage currently, because you still need a way to call Spack:
- always using the binary (uberenv-libs/spack/bin/spack) ?
- loading spack shell setup : . ubernev-libs/spack/share/spack/setup.sh ?

Reply:
@cyrush -- maybe we could generate a small script ``run_spack`` in the ``uberenv_libs`` directory that runs spack with the desired `persistent` parameters.

This is similar to what ``axom`` does in its ``config_build`` script to use the same version of cmake that was used to configure axom:
https://github.com/LLNL/axom/blob/61cd97c5aad8c96fd5d3defe16f5f72667ee1b99/config-build.py#L217-L230

A side benefit would be that you no longer need to run ``spack/bin/spack``

Reply:
After a bit more thought, I think it is crucial to export a `run_spack` script (and document this) to help users know how to setup their spack environment, as per my previous comment.